types: get rid of linearization in deserialize()

michoecho opened the pull request:
Citing #6138:
> In the past few years we have converted most of our codebase to work in terms of fragmented buffers, instead of linearised ones, to help avoid large allocations that put large pressure on the memory allocator.
> One prominent component that still works exclusively in terms of linearised buffers is the types hierarchy, more specifically the de/serialization code to/from CQL format. Note that for most types, this is the same as our internal format, notable exceptions are non-frozen collections and user types.
> 
> Most types are expected to contain reasonably small values, but texts, blobs and especially collections can get very large. Since the entire hierarchy shares a common interface we can either transition all or none to work with fragmented buffers.

This series gets rid of intermediate linearizations in deserialization. The next steps are removing linearizations from serialization, validation and comparison code.

Series summary:
- Fix a bug in `fragmented_temporary_buffer::view::remove_prefix`. (Discovered while testing. Since it wasn't discovered earlier, I guess it doesn't occur in any code path in master.)
- Add a `FragmentedView` concept to allow uniform handling of various types of fragmented buffers (`bytes_view`, `temporary_fragmented_buffer::view`, `ser::buffer_view` and likely `managed_bytes_view` in the future).
- Implement `FragmentedView` for relevant fragmented buffer types.
- Add helper functions for reading from `FragmentedView`.
- Switch `deserialize()` and all its helpers from `bytes_view` to `FragmentedView`.
- Remove `with_linearized()` calls which just became unnecessary.
- Add an optimization for single-fragment cases.

The addition of `FragmentedView` might be controversial, because another concept meant for the same purpose - `FragmentRange` - is already used. Unfortunately, it lacks the functionality we need. The main (only?) thing we want to do with a fragmented buffer is to extract a prefix from it and `FragmentRange` gives us no way to do that, because it's immutable by design. We can work around that by wrapping it into a mutable view which will track the offset into the immutable `FragmentRange`, and that's exactly what `linearizing_input_stream` is. But it's wasteful. `linearizing_input_stream` is a heavy type, unsuitable for passing around as a view - it stores a pair of fragment iterators, a fragment view and a size (11 words) to conform to the iterator-based design of `FragmentRange`, when one fragment iterator (4 words) already contains all needed state, just hidden.
I suggest we replace `FragmentRange` with `FragmentedView` (or something similar) altogether.

Refs: #6138

avikivity commented on utils/fragment_range.hh:
```diff
@@ -173,3 +173,28 @@ concept FragmentedMutableView = requires (T view) {
+template<FragmentedView View, typename Function>
+requires (!FragmentRange<View>) && requires (Function fn, bytes_view bv) {
+    fn(bv);
+}
```
You can write the latter constraint as std::invocable<Function, bytes_view>

michoecho replied:
Done.

avikivity commented on types.cc:
```diff
@@ -1883,7 +1883,8 @@ static uint32_t deserialize_value(const simple_date_type_impl&, bytes_view v) {
 
-static int64_t deserialize_value(const time_type_impl&, bytes_view v) {
+template<FragmentedView View>
+int64_t deserialize_value(const time_type_impl&, View v) {
```
You can use the abbreviated syntax: `int64_t deserialize_value(const time_type_impl&, FragmentedView auto v)`, if you like.

michoecho replied:
Wow, yet another concept syntax. I didn't know that was a thing.

> If you like

 I don't know. Should I like?

avikivity replied:
> Should I like?

Good question. One the one hand, it has less fluff. On the other hand, it hides the fact that it is a template. On the feet, we may learn to recognize the syntax in time, and perhaps the fact that it is a template is not so important.

avikivity commented on types.cc:
```diff
@@ -1811,16 +1811,20 @@ static data_value deserialize_aux(const tuple_type_impl& t, bytes_view v) {
-    auto negative = v.front() < 0;
+template<FragmentedView View>
+utils::multiprecision_int deserialize_value(const varint_type_impl&, View v) {
+    bool negative = v.current_fragment().front() < 0;
```
What if the view is empty? The existing code will survive it.

I don't think we can actually get called here with an empty view, but better to be prepared.

michoecho replied:
> The existing code will survive it.

What do you mean? Calling `front()` on an empty `bytes_view` (aka `std::basic_string_view`) is undefined behaviour.

avikivity replied:
You are right, I withdraw the comment.

avikivity commented on cql3/maps.cc:
```diff
@@ -158,15 +158,13 @@ maps::value::from_serialized(const fragmented_temporary_buffer::view& fragmented
         // FIXME: deserialize_for_native_protocol?!
-      return with_linearized(fragmented_value, [&] (bytes_view value) {
-        auto m = value_cast<map_type_impl::native_type>(type.deserialize(value, sf));
+        auto m = value_cast<map_type_impl::native_type>(type.deserialize(fragmented_value, sf));
```
In many cases, the value will come in a single fragment (e.g. someone passed us an int32, the odds of it being fragmented in our 8k buffers are 1:2000). But we will pay the cost of fragmented deserialization nevertheless (endless checking).

So abstract_type::deserialize() can check whether we have 0 or 1 fragments, and call the bytes_view variant in those cases.

Or maybe it can be a helper function that takes in a fragmented view and a function, checks if it has more than one fragment, and calls the inner function with a single_fragment_view if possible, and with the original input if not. This may replace many checks during inner processing with just one check in the outer processing.

This can come post merge.

denesb replied:
I think long-term we should examine why do we need these serialize-deserialize methods at all. We need to be able to do 3 things with our values:
* de/serialize to CQL wire format: for most types this is the same as the in-memory format
* de/serialize to on-disk format: I suspect for most types the same applies as before
* compare types

Neither of these should require deserializing a fragmented buffer belonging to a UTF8 value into a `sstring` C++ type, then re-serializing it into another format or comparing it to another `sstring`. We should drop the intermediate format.

One more uses case is pretty printing and genuine conversion to C++ types in tests. Neither of these is on any hot path.

michoecho replied:
@avikivity Done in c990869. Please take a look.

avikivity commented on cql3/maps.cc:
```diff
@@ -158,15 +158,13 @@ maps::value::from_serialized(const fragmented_temporary_buffer::view& fragmented
         // but compose does the validation (so we're fine).
         // FIXME: deserialize_for_native_protocol?!
-      return with_linearized(fragmented_value, [&] (bytes_view value) {
-        auto m = value_cast<map_type_impl::native_type>(type.deserialize(value, sf));
```
It pays to be lazy - we did not fix the half-indent, and now we don't need to.

avikivity commented on types.cc:
```diff
@@ -1822,6 +1822,29 @@ static void serialize(const abstract_type& t, const void* value, bytes::iterator
+    struct visitor {
+        View v;
+        cql_serialization_format sf;
+        data_value operator()(const abstract_type&) { abort(); }
```
Use on_internal_error() so as not to upgrade a but to total loss of availability.

michoecho replied:
I will do that, but this code is *very* unreachable (a new collection type would need to be added) so it doesn't really matter.

avikivity commented on types.cc:
```diff
@@ -1822,6 +1822,29 @@ static void serialize(const abstract_type& t, const void* value, bytes::iterator
+        }
+        data_value operator()(const set_type_impl& t) {
+            return t.deserialize(v, sf);
+        }
```
Why do you need these three special cases? Can't you write `[sf, v] (const auto& t) { return t.deserialize(v, sf); }`?

Looks like the only thing the visitor does apart from that is to abort on a bad type.

michoecho replied:
Collections have a two-argument version of deserialize (`collection_type_impl::deserialize`) which allows specifying a `cql_serialization_format`. This used to be a virtual method, but since we want to template it, we need to make it non-virtual and dispatch the call manually, with an abstract_type visitor. Just like we do in `abstract_type::deserialize`. (We could also write a `switch` over the types ourselves. Maybe we could help the compiler with optimization that way). 

You can't `[sf, v] (const auto& t) { return t.deserialize(v, sf); }` because a visitor needs to match all types, and non-collection types don't have a `deserialize` with two arguments.

avikivity replied:
I see. Maybe they should have the two-argument form. But even if we agree they should, let's leave it for later.

avikivity commented:
Looks very good, very easy to review. I left some minor comments.

michoecho commented on utils/fragment_range.hh:
```diff
@@ -198,3 +198,16 @@ decltype(auto) with_linearized(const View& v, Function&& fn)
+    bytes_view _view;
+public:
+    using fragment_type = bytes_view;
+    explicit single_fragmented_view(bytes_view bv) : _view(bv) {}
```
I made it explicit specifically so that it isn't seamless. It will make it easier to find potential linearizations and/or uses of old `bytes_view` versions of stuff.

michoecho commented on types.cc:
```diff
@@ -1899,9 +1899,24 @@ static T deserialize_value(const integer_type_impl<T>& t, bytes_view v) {
+sstring deserialize_value(const string_type_impl&, View v) {
     // FIXME: validation?
-    return sstring(reinterpret_cast<const char*>(v.begin()), v.size());
+    sstring buf(sstring::initialized_later(), v.size_bytes());
```
True. But it's an independent issue for another pull request.

michoecho commented on utils/fragment_range.hh:
```diff
@@ -173,3 +173,28 @@ concept FragmentedMutableView = requires (T view) {
+    if (v.size_bytes() == v.current_fragment().size()) [[likely]] {
+        return fn(v.current_fragment());
+    } else {
+        return fn(linearized(v));
```
I don't think so. In C++, temporaries stay alive until the full containing expression is evaluated, no?

denesb replied:
I don't think so. Here, the compiler has no way to know that `bytes_view` depends on the lifetime of `bytes`. I think only the value immediately passed to the function is lifetime-extended. But an experiment should be easy enough to confirm/deny.

michoecho replied:
After https://en.cppreference.com/w/cpp/language/lifetime:
> All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created, and if multiple temporary objects were created, they are destroyed in the order opposite to the order of creation.

In other words, for:
```
struct A {
    A() { std::cout << "A "; }
    ~A() { std::cout << "~A "; }
};
struct B {
    B() { std::cout << "B "; }
    ~B() { std::cout << "~B "; }
    operator A() { return A(); }
};
void f(A) {
    std::cout << "f "; 
}
int main() { f(B()); }
```
the only possible output is `B A f ~A ~B `.

michoecho replied:
```
int& f(int&& x) { return &x; }
int main() {
    int *a;
    a = &f(0);
    std::cout << *a;
}
```
is undefined behaviour, but
```
int& f(int&& x) { return &x; }
int main() {
    int *a;
    a = &f(0),
    std::cout << *a;
}
```
is perfectly good C++.

denesb replied:
We are safe then. I learned something new today.

michoecho commented:
Post-review changes:
- Adressed the minor comments in https://github.com/scylladb/scylla/pull/7692#discussion_r530246766, https://github.com/scylladb/scylla/pull/7692#discussion_r530266440 and https://github.com/scylladb/scylla/pull/7692#discussion_r530880687.
- Corrected a comment.
- Documented the preconditions of FragmentedView methods.
- Added an optimization for single-fragment views (c990869), as suggested in https://github.com/scylladb/scylla/pull/7692#discussion_r530259586.

michoecho commented on types.hh:
```diff
@@ -1108,6 +1108,22 @@ typename Type::value_type deserialize_value(Type& t, bytes_view v) {
 
+template<FragmentedView View>
+void read_fragmented(View& v, size_t n, bytes::value_type* out) {
+    assert(n <= v.size_bytes());
```
Done.

michoecho commented:
I have removed linearizations from validate() (https://github.com/michoecho/scylla/compare/types_remove_linearization...michoecho:types_remove_linearization_2) and moved on to serialize(). I wonder, should I push the validate() changes here for review, or make a separate PR?

avikivity commented:
@michoecho this one is fat enough, let's get it in

avikivity commented:
@denesb if you have no further comments, I can merge.

michoecho commented:
There's a lot of very small commits here. Should I do some squashing before merge?

avikivity commented:
I love small commits!

michoecho commented:
New change: I forgot to push the single-fragment optimization for the collection (two-argument) form of deserialize() earlier. It's the same optimization as for the single-argument form.

denesb commented:
>     * Added an optimization for single-fragment views ([c990869](https://github.com/scylladb/scylla/commit/c990869c8ea66e24c96d978bfc1c38fc1c74b3b2)), as suggested in [#7692 (comment)](https://github.com/scylladb/scylla/pull/7692#discussion_r530259586).

I don't see how this resolves the problem. All the different overloads, including the `single_fragment_view` uses the same generic implementation, I didn't find any `single_fragment_view` specific overload taking advantage of the fact that we have single fragment. I can see that we have `bytes_view` variants of the various `read_*` methods, but I don't see when a `single_fragment_view` is converted to `bytes_view`. What am I missing?

michoecho commented:
@denesb It's not about `read_*`, but about the `FragmentedView` methods. `prefix()`, `remove_prefix()` and `remove_current()` are trivial for `single_fragmented_view`, but complex for other `FragmentedView` types. That's why it's an optimization.
`single_fragmented_view` _is_ a `bytes_view` underneath, so `read_*` functions don't do any extra work compared to their `bytes_view` counterparts. (With the exception of `read_fragmented`, which I used to read IPv6 addresses, which previously were read by simple `memcpy`. It probably deserves a special overload. I will add it in a moment.)
Besides that, we are now passing 2 words (as opposed to 4 words of multi-fragment views) through optimization barriers.

denesb commented:
I see, thanks for explaining.

denesb reviewed (approved):
LGTM.

michoecho commented:
Minor changes: https://github.com/scylladb/scylla/compare/4953df5..588bebb
- Removed the bounds check from `read_fragmented`, because it's always called after they have already been checked. Added a comment documenting that.
- Made `read_simple_exactly` not modify its argument, to match the behaviour of the `bytes_view` version exactly (though it doesn't matter).
- Added a specialization for `read_fragmented(single_fragmented_view)`, because the generic version does more work than necessary for single-fragment views.

michoecho commented:
Fixed a linkage mistake I just made: https://github.com/scylladb/scylla/compare/588bebb..e504bba

michoecho commented:
@avikivity I haven't modified the PR description (which has just become the merge commit message in next) after review fixes, so it contains outdated commit hashes. Is that fine?

avikivity commented:
It's not the end of the world. In the future avoid using commit hashes in descriptions, they're too volatile until actually committed.

michoecho commented:
Added a missing explicit instantiation for `collection_type_impl::deserialize(single_fragmented_view) and collection_type_impl::deserialize(temporary_fragmented_buffer::view)`. (Its lack caused `release` builds to fail, because the implicit instantiations were optimized away.)
Removed commit hashes from the PR description.
